### PR TITLE
feat/quan 12 implement google ads on llc

### PR DESCRIPTION
**Pull request by kherma**

### CODE REVIEW

**Comment by bskinn** [comment]
@kherma I could see the TypeScript constraints here causing problems. 

We *don't* want any of the gtag scripts running on any of the Labs pages...only the Consulting pages.

But, since the `Form` is a shared component, `gtag_report_conversion()` will (i) have to be correctly typed for both sites, but (ii) only be called on Consulting.

I came across a site that had instructions for getting past this sort of typing issue with gtag when I was investigating this for myself... I can try to find it again if that would be helpful.

File: libs/shared/ui-components/src/Form/Form.tsx [hunk 1]
```diff
@@ -35,6 +36,7 @@ export const Form: FC<TFormProps> = (props) => {
   };
 
   const onSubmit = handleSubmit((formValues): void => {
+    gtag_report_conversion();
```
Comment (bskinn) [inline libs/shared/ui-components/src/Form/Form.tsx:39]: Pretty sure the function is expecting to receive the current page URL here. Not sure if it's as simple as adding `window.location.href`?

```suggestion
    gtag_report_conversion(window.location.href);
```

**Comment by bskinn** [comment]
I looked at the source for a page of the preview build, and the gtag code _does_ show in one of the webpack bundles. :+1:

Per @bsodenkamp, though, it sounds like we won't be able to test Ads's conversion pipeline until the code is live at `quansight.com`. So, we're going to have to just push to live and debug from there.

I did go ahead and submit a test contact form using the preview build on this PR, and it made it successfully into Pipedrive. So, even if there are glitches to iron out with the Ads code, our lead collection machinery should not be affected.

**Comment by bskinn** [comment]
@gabalafou, can you give this a review today? We'd like it to go live with tomorrow's build. Thanks!

File: apps/consulting/pages/_app.tsx [hunk 2]
```diff
@@ -12,6 +12,17 @@ function CustomApp({ Component, pageProps }: AppProps): JSX.Element {
...
 
+      <Script
+        strategy="afterInteractive"
+        src={`https://www.googletagmanager.com/gtag/js?id=${process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS}`}
+      />
+      <Script id="google-analytics-script" strategy="afterInteractive">
+        {`window.dataLayer = window.dataLayer || [];
+        function gtag(){dataLayer.push(arguments);}
+        gtag('js', new Date());
+        gtag('config', '${process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS}');`}
```
Comment (gabalafou) [inline apps/consulting/pages/_app.tsx:23]: Could we rewrite this as:

```js
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS);
```

? I don't see the need to quote the entire code snippet.

File: libs/shared/utils/src/gtag/reportConversion.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+export const gtag_report_conversion = (url?: Location | undefined) => {
+  const callback = function () {
+    if (typeof url != 'undefined') {
+      window.location = url;
+    }
+  };
+
+  window.gtag('event', 'conversion', {
+    send_to: `${process.env['NEXT_PUBLIC_GOOGLE_ANALYTICS']}/${process.env['NEXT_PUBLIC_CONVERSION_LABEL']}`,
```
Comment (gabalafou) [inline libs/shared/utils/src/gtag/reportConversion.ts:9]: I'm not sure it makes any difference but is there any reason you're doing `process.env.['NEXT_PUBLIC_GOOGLE_ANALYTICS']` vs `process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS`?

**Review by gabalafou — COMMENTED** [review]
Got some questions for you :)

File: apps/consulting/index.d.ts [hunk 4]
```diff
@@ -1,6 +1,12 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
 declare module '*.svg' {
-  const content: any
-  export const ReactComponent: any
-  export default content
+  const content: any;
+  export const ReactComponent: any;
+  export default content;
+
+  declare global {
```
[the remark this answers is not in the document]
Comment (bskinn) [reply apps/consulting/index.d.ts:7]: @kherma please reply

File: libs/shared/utils/src/gtag/reportConversion.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    event_callback: callback,
+  });
+  return false;
```
[the remark this answers is not in the document]
Comment (bskinn) [reply libs/shared/utils/src/gtag/reportConversion.ts:12]: @kherma please reply

File: apps/consulting/pages/_app.tsx (hunk 2 above)
Comment (bskinn) [reply apps/consulting/pages/_app.tsx:23]: @kherma can probably speak to this better than I can, but when I was trying earlier to implement this myself, whenever it wasn't quoted, Typescript (I think it was) complained vociferously.

File: libs/shared/utils/src/gtag/reportConversion.ts (hunk 3 above)
Comment (bskinn) [reply libs/shared/utils/src/gtag/reportConversion.ts:9]: @kherma

File: libs/shared/ui-components/src/Form/Form.tsx [hunk 6]
```diff
@@ -35,6 +36,7 @@ export const Form: FC<TFormProps> = (props) => {
   };
 
   const onSubmit = handleSubmit((formValues): void => {
+    gtag_report_conversion(window?.location);
```
[the remark this answers is not in the document]
Comment (gabalafou) [reply libs/shared/ui-components/src/Form/Form.tsx:39]: I should have been more specific. I don't understand why there's a question mark after `window?`. Is this to make TypeScript happy? What's going on here? I believe the `window` object should always be available by the time the submit handler is called.

File: libs/shared/utils/src/gtag/reportConversion.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+export const gtag_report_conversion = (url?: Location | undefined) => {
+  const callback = function () {
```
[the remark this answers is not in the document]
Comment (gabalafou) [reply libs/shared/utils/src/gtag/reportConversion.ts:2]: I see. But then my question would be something more along the lines of: are we sure we're using this snippet correctly? 

Also, can we put a comment at the top of this file explaining that the code was copy-pasted, and provide a link or instructions where to get the code snippet?

File: apps/consulting/pages/_app.tsx (hunk 2 above)
Comment (gabalafou) [reply apps/consulting/pages/_app.tsx:23]: Oh I see, [it's a Next.js thing](https://nextjs.org/docs/basic-features/script#inline-scripts).

File: apps/consulting/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Katebrack) [reply apps/consulting/index.d.ts:7]: @kherma adding another follow-up here.

File: libs/shared/utils/src/gtag/reportConversion.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Katebrack) [reply libs/shared/utils/src/gtag/reportConversion.ts:12]: @kherma here as well.

File: libs/shared/utils/src/gtag/reportConversion.ts (hunk 3 above)
Comment (bskinn) [reply libs/shared/utils/src/gtag/reportConversion.ts:9]: This is probably where the `undefined`s are coming from. I'm going to remove the quotes and see if that fixes it.

Comment (bskinn) [reply libs/shared/utils/src/gtag/reportConversion.ts:9]: Check that: I'm going to add these env variables to Vercel, and see if that fixes it.

File: apps/consulting/pages/_app.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (kherma) [reply apps/consulting/pages/_app.tsx:23]: @gabalafou  It's more typescript thing: https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature Can also be disabled in config.

File: libs/shared/utils/src/gtag/reportConversion.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kherma) [reply libs/shared/utils/src/gtag/reportConversion.ts:4]: @gabalafou Snippet from https://github.com/Quansight/Quansight-website/issues/404#issuecomment-1248287234

**Review by gabalafou — APPROVED** [review]

**Comment by gabalafou** [comment]
I'm still not sure about that callback, but I guess we'll just deploy and see?

**Comment by gabalafou** [comment]
That's not what I see:

```
gtag_report_conversion(window.location);
```

🤔

**Comment by bskinn** [comment]
Indeed, I think my change got reverted when @kherma removed the `?` in a fix commit.

@kherma, I was wrong in our earlier discussion, we want ***no argument*** passed to the `gtag_report_conversion()` function. I'm going to add a commit to your branch shortly to make this change.

**Comment by bskinn** [comment]
Fixed. **Thank you** for catching this, @gabalafou!

File: libs/shared/ui-components/src/Form/Form.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (bskinn) [reply libs/shared/ui-components/src/Form/Form.tsx]: The argument to `gtag_report_conversion` has been removed entirely, and thus the `window` vs `window?` is moot, so I'm going to resolve this conversation --- further followup can take place outside this repo.

File: apps/consulting/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bskinn) [reply apps/consulting/index.d.ts]: The global declaration was moved outside the `module` declaration in f1586433.

File: libs/shared/utils/src/gtag/reportConversion.ts (hunk 3 above)
Comment (bskinn) [reply libs/shared/utils/src/gtag/reportConversion.ts:11]: This implementation choice doesn't affect this PR's behavior, so I'm going to resolve it here. Further discussion can take place outside the repo.

**Comment by bskinn** [comment]
I just re-confirmed that the conversion reporting works correctly on the most recent preview build, after re-removing the argument to the call to `gtag_report_conversion()`, so I'm going to go ahead and merge.

**Comment by bskinn** [comment]
You can go ahead and change status this to completed status in Jira, @kherma. Thank you for the implementation!

**Pull request merged**